README.md
# stereo-vision

| License | Versioning | Build |
| ------- | ---------- | ----- |
| [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT) | [![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release) | [![Build status](https://ci.appveyor.com/api/projects/status/cilnfjwa52c0bllu/branch/master?svg=true)](https://ci.appveyor.com/project/nikAizuddin/stereo-vision/branch/master) |


Stereo Vision project.


## Podman

Build image:
```
podman build -t extra2000/stereovision .
```


### Example usage

Create output directory:
```
mkdir -pv results
```

For SELinux, label work files as `container_file_t` to allow to be mounted into container:
```
chcon -v -t container_file_t left-image.jpg right-image.jpg out.jpg results
```

How to run:
```
podman run -it --rm -v ./left-image.jpg:/opt/workdir/left-image.jpg:ro -v ./right-image.jpg:/opt/workdir/right-image.jpg:ro -v ./results:/opt/workdir/results:rw localhost/extra2000/stereovision stereovision generate-disparity-image left-image.jpg right-image.jpg results/out.jpg
```


## PIP

Installations:
```
python3 -m pip install .
```

Example usage:
```
stereovision generate-disparity-image left-image.jpg right-image.jpg out.jpg
```


## How to Calculate Camera Matrix Focal Length?

Let,
  g = camera matrix focal length in unit pixel
  f = optical focal length in unit mm

Assume a Pixel 4A phone camera which has 1.40 micro meter pixel pitch and 4.38 mm focal length. The value of g would be:

g = (4.38 * 1000) / 1.40
  = ~3128.6 pixels

Note: 1mm = 1000 micro meter. The optical focal length f needed to be converted to micro meter.


## How to find out pixel pitch for Blender

scale = scene.render.resolution_percentage / 100
pixels_in_u_per_mm = resolution_x_in_px * scale / sensor_width_in_mm
pixels_in_v_per_mm = resolution_y_in_px * scale * aspect_ratio / sensor_height_in_mm
pixel_size_in_u_direction = 1/pixels_in_u_per_mm
pixel_size_in_v_direction = 1/pixels_in_v_per_mm


CHANGELOG.md
# Changelog

## [1.1.0](https://github.com/extra2000/stereo-vision/compare/v1.0.0...v1.1.0) (2022-06-08)


### Features

* add point cloud ([fc598ef](https://github.com/extra2000/stereo-vision/commit/fc598efb6af8ae49bba714bc2ebc44d2dd07a926))
* **disparity:** add `--win-size` argument ([ec02c3d](https://github.com/extra2000/stereo-vision/commit/ec02c3d6d45bfe29ac676745c8ffa88abf8bd803))


### Performance Improvements

* downscale image before compute Stereo SGBM ([44f64d9](https://github.com/extra2000/stereo-vision/commit/44f64d9d6c304f8b4b616c4c42499864402c9950))


### Code Refactoring

* **disparity:** replace `num_disparities` with `max_disparity` and adjust default disparity range ([a006d85](https://github.com/extra2000/stereo-vision/commit/a006d85d0cf9e95f65b05e804d7c1785bb6bdf69))
* **imread:** read images as grayscale ([806dc5a](https://github.com/extra2000/stereo-vision/commit/806dc5a386c0c1658cc65cdd31699f61501bfc3d))

## 1.0.0 (2021-09-29)


### Features

* initial commit ([fbf3931](https://github.com/extra2000/stereo-vision/commit/fbf39317cf11387fda69aa5b5e6fdd39c297b42a))


### Documentations

* **README:** update `README.md` ([7d85fa3](https://github.com/extra2000/stereo-vision/commit/7d85fa31df4629205bcbf1bf26dccacb94a90cb0))


### Continuous Integrations

* add AppVeyor and `semantic-release` ([8788494](https://github.com/extra2000/stereo-vision/commit/878849466d7cec09391d6ba7e3cff662089952ee))


stereo_vision/stereo_vision.py
import os
import logging

import numpy as np
import cv2

logger = logging.getLogger(__name__)


class StereoVision:
    """Stereo Vision
    """

    def __init__(self):
        """Implement Stereo Vision
        """
        self.cfg = None


    @staticmethod
    def get_disparity_map(leftimage, rightimage, **kwargs):
        """
        Combine stereo image into a single image which represents disparity map.

        Parameters
        ----------
        leftimage : str
            Left image filename.
        rightimage : str
            Right image filename.
        min_disparity : int
            Minimum disparity. Default is -16.
        max_disparity : int
            Number of disparities. Default is 16.
        win_size : int
            Window size. Default is 2.
        block_size : int
            Block size. Default is 1.
        disp12_max_diff : int
            Maximum allowed difference. Default is -1.
        uniqueness_ratio : int
            Uniqueness ratio. Default is 15.
        speckle_window_size : int
            Speckle window size. Default is 200.
        speckle_range : int
            Speckle range. Default is 64.
        sgbm_mode : str
            SGBM mode. Available choices are 'hh', '3way', or 'sgbm'. Default is '3way'.

        Returns
        -------
        cv2.image
            Disparity map with the same dimension as input image.
        """

        img_left = cv2.pyrDown(cv2.imread(leftimage, cv2.IMREAD_GRAYSCALE))
        img_right = cv2.pyrDown(cv2.imread(rightimage, cv2.IMREAD_GRAYSCALE))

        # Setting parameters for StereoSGBM algorithm
        min_disparity = kwargs.get('min_disparity', -16)
        max_disparity = kwargs.get('max_disparity', 16)
        win_size = kwargs.get('win_size', 2)
        block_size = kwargs.get('block_size', 1)
        disp12_max_diff = kwargs.get('disp12_max_diff', -1)
        uniqueness_ratio = kwargs.get('uniqueness_ratio', 15)
        speckle_window_size = kwargs.get('speckle_window_size', 200)
        speckle_range = kwargs.get('speckle_range', 64)

        param_sgbm_mode = kwargs.get('sgbm_mode', '3way')

        if param_sgbm_mode == 'hh':
            sgbm_mode = cv2.STEREO_SGBM_MODE_HH
        elif param_sgbm_mode == '3way':
            sgbm_mode = cv2.STEREO_SGBM_MODE_SGBM_3WAY
        elif param_sgbm_mode == 'sgbm':
            sgbm_mode = cv2.STEREO_SGBM_MODE_SGBM
        else:
            raise ValueError(param_sgbm_mode)

        num_disparities = max_disparity - min_disparity

        # Creating an object of StereoSGBM algorithm
        stereo = cv2.StereoSGBM_create(
            minDisparity=min_disparity,
            numDisparities=num_disparities,
            blockSize=block_size,
            disp12MaxDiff=disp12_max_diff,
            uniquenessRatio=uniqueness_ratio,
            speckleWindowSize=speckle_window_size,
            speckleRange=speckle_range,
            P1=8*3*win_size**2,
            P2=32*3*win_size**2,
            mode=sgbm_mode
        )

        # Calculating disparity using the StereoSGBM algorithm
        disparity_map = stereo.compute(img_left, img_right).astype(np.float32)

        return cv2.pyrUp(disparity_map)


    @staticmethod
    def get_point_cloud(leftimage, disparity_map, **kwargs):
        """
        Generate point cloud based on disparity map

        Parameters
        ----------
        leftimage : str
            Left image filename.
        disparity_map : cv2.image
            Disparity map

        Returns
        -------
        np.array
            Point cloud data.
        np.array
            Colors for the point cloud.
        """

        colors = cv2.imread(leftimage, cv2.COLOR_BGR2RGB)

        h, w = disparity_map.shape[:2]
        f = (50 * 1000) / 16.7
        Q = np.float32([
            [1,  0, 0, -0.5*w],
            [0, -1, 0,  0.5*h],
            [0,  0, 0, -f],
            [0,  0, 1,  0]
        ])
        points = cv2.reprojectImageTo3D(disparity_map, Q)
        mask = disparity_map > disparity_map.min()
        out_points = points[mask]
        out_colors = colors[mask]

        return out_points, out_colors


    @staticmethod
    def write_disparity_map(filename, outimage):
        """Write disparity map to image file

        Parameters
        ----------
        filename : str
            Filename to write output image.
        outimage : cv2.image
            Output image.
        """

        normimg = np.abs(cv2.normalize(outimage, 0, 255, cv2.NORM_MINMAX)*255).astype(np.uint8)
        cv2.imwrite(filename, normimg)


    @staticmethod
    def write_point_cloud(plyfile, points, colors):
        """Write point clouds to PLY file

        Parameters
        ----------
        plyfile : str
            PLY filename
        points : np.array
            Points.
        colors : np.array
            Colors for the points.
        """

        verts = points.reshape(-1, 3)
        colors = colors.reshape(-1, 3)
        verts = np.hstack([verts, colors])
        with open(plyfile, 'wb') as f:
            # write PLY header
            f.write(('ply\n').encode('utf-8'))
            f.write(('format ascii 1.0\n').encode('utf-8'))
            f.write(('element vertex {}\n'.format(len(verts))).encode('utf-8'))
            f.write(('property float x\n').encode('utf-8'))
            f.write(('property float y\n').encode('utf-8'))
            f.write(('property float z\n').encode('utf-8'))
            f.write(('property uchar red\n').encode('utf-8'))
            f.write(('property uchar green\n').encode('utf-8'))
            f.write(('property uchar blue\n').encode('utf-8'))
            f.write(('end_header\n').encode('utf-8'))
            # write point clouds including colors
            np.savetxt(f, verts, fmt='%f %f %f %d %d %d ')


stereo_vision/console/main.py
"""Entry-point for stereovision executable.
"""

import os
import sys
import logging
import logging.handlers
import argparse
from functools import partial

from stereo_vision import StereoVision, __version__

logger = logging.getLogger(__name__)


def main():
    try:
        _app()
    except KeyboardInterrupt as e:
        logger.exception('Received keyboard interrupt. Exiting.', exc_info=False)
    except Exception as e:
        logger.error('Exception type: {}'.format(type(e)))
        logger.error(str(e))
        logger.exception(str(e), exc_info=True)
        sys.exit(1)


def _app():
    args, parser = _parse_args()

    _init_logger()

    logger.info('Stereo Vision CLI version {}'.format(__version__))

    stereovision = StereoVision()

    command = {
        'generate-disparity-image': partial(_get_disparity_map, stereovision, args),
        'generate-point-cloud': partial(_get_point_cloud, stereovision, args),
    }

    choice = command.get(args.command, lambda: parser.print_help())
    choice()


def _get_disparity_map(stereovision, args):
    disparity_map = stereovision.get_disparity_map(
        args.leftimage,
        args.rightimage,
        min_disparity=vars(args)['min_disparity'],
        max_disparity=vars(args)['max_disparity'],
        win_size=vars(args)['win_size'],
        block_size=vars(args)['block_size'],
        disp12_max_diff=vars(args)['disp12_max_diff'],
        uniqueness_ratio=vars(args)['uniqueness_ratio'],
        speckle_window_size=vars(args)['speckle_window_size'],
        speckle_range=vars(args)['speckle_range'],
        sgbm_mode=vars(args)['sgbm_mode']
    )
    stereovision.write_disparity_map(args.outimage, disparity_map)


def _get_point_cloud(stereovision, args):
    disparity_map = stereovision.get_disparity_map(
        args.leftimage,
        args.rightimage,
        min_disparity=vars(args)['min_disparity'],
        max_disparity=vars(args)['max_disparity'],
        win_size=vars(args)['win_size'],
        block_size=vars(args)['block_size'],
        disp12_max_diff=vars(args)['disp12_max_diff'],
        uniqueness_ratio=vars(args)['uniqueness_ratio'],
        speckle_window_size=vars(args)['speckle_window_size'],
        speckle_range=vars(args)['speckle_range'],
        sgbm_mode=vars(args)['sgbm_mode']
    )
    points, colors = stereovision.get_point_cloud(args.leftimage, disparity_map)
    stereovision.write_point_cloud(vars(args)['plyfile'], points, colors)


def _init_logger():
    formatter = logging.Formatter('%(asctime)s %(process)d:%(thread)d:%(levelname)s:%(name)s:%(lineno)d: %(message)s')

    stream_handler = logging.StreamHandler()
    stream_handler.setFormatter(formatter)
    stream_handler.setLevel('INFO')

    rootlogger = logging.getLogger('stereo_vision')
    rootlogger.addHandler(stream_handler)
    rootlogger.setLevel('DEBUG')


def _parse_args():
    parser = argparse.ArgumentParser(
        description='Stereo Vision CLI version {}.'.format(__version__),
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )

    subparser = parser.add_subparsers(help='CLI command', dest='command')

    get_disp_img = subparser.add_parser('generate-disparity-image', help='Combine stereo image into a single image which represents disparity map')
    get_disp_img.add_argument('--min-disparity', type=int, help='Minimum disparity. Default is %(default)d', default=-16)
    get_disp_img.add_argument('--max-disparity', type=int, help='Maximum disparity. Default is %(default)d', default=16)
    get_disp_img.add_argument('--win-size', type=int, help='Window size. Default is %(default)d', default=2)
    get_disp_img.add_argument('--block-size', type=int, help='Block size. Default is %(default)d', default=1)
    get_disp_img.add_argument('--disp12-max-diff', type=int, help='Maximum allowed difference. Default is %(default)d', default=-1)
    get_disp_img.add_argument('--uniqueness-ratio', type=int, help='Uniqueness ratio. Default is %(default)d', default=15)
    get_disp_img.add_argument('--speckle-window-size', type=int, help='Speckle window size. Default is %(default)d', default=200)
    get_disp_img.add_argument('--speckle-range', type=int, help='Speckle range. Default is %(default)d', default=64)
    get_disp_img.add_argument('--sgbm-mode', type=str, help='SGBM mode. Default is %(default)s', choices=['hh', '3way', 'sgbm'], default='3way')
    get_disp_img.add_argument('leftimage', type=str, help='Left image')
    get_disp_img.add_argument('rightimage', type=str, help='Right image')
    get_disp_img.add_argument('outimage', type=str, help='Output image')

    get_pntcloud_img = subparser.add_parser('generate-point-cloud', help='Compute disparity map and generate point cloud')
    get_pntcloud_img.add_argument('--min-disparity', type=int, help='Minimum disparity. Default is %(default)d', default=-16)
    get_pntcloud_img.add_argument('--max-disparity', type=int, help='Maximum disparity. Default is %(default)d', default=16)
    get_pntcloud_img.add_argument('--win-size', type=int, help='Window size. Default is %(default)d', default=2)
    get_pntcloud_img.add_argument('--block-size', type=int, help='Block size. Default is %(default)d', default=1)
    get_pntcloud_img.add_argument('--disp12-max-diff', type=int, help='Maximum allowed difference. Default is %(default)d', default=-1)
    get_pntcloud_img.add_argument('--uniqueness-ratio', type=int, help='Uniqueness ratio. Default is %(default)d', default=15)
    get_pntcloud_img.add_argument('--speckle-window-size', type=int, help='Speckle window size. Default is %(default)d', default=200)
    get_pntcloud_img.add_argument('--speckle-range', type=int, help='Speckle range. Default is %(default)d', default=64)
    get_pntcloud_img.add_argument('--sgbm-mode', type=str, help='SGBM mode. Default is %(default)s', choices=['hh', '3way', 'sgbm'], default='3way')
    get_pntcloud_img.add_argument('leftimage', type=str, help='Left image')
    get_pntcloud_img.add_argument('rightimage', type=str, help='Right image')
    get_pntcloud_img.add_argument('plyfile', type=str, help='Output PLY file')

    args = parser.parse_args()

    return args, parser
